File: Android/virtues_dynamic/src/main/java/com/appdimens/dynamic/code/AppDimensDynamic.kt
/**
 * Author & Developer: Jean Bodenberg
 * GIT: https://github.com/www-virtues-ag/virtues-dimens.git
 * Date: 2025-10-04
 *
 * Library: Virtues
 *
 * Description:
 * The Virtues library is a dimension management system that automatically
 * adjusts Dp, Sp, and Px values in a responsive and mathematically refined way,
 * ensuring layout consistency across any screen size or ratio.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ag.virtues.dimens.dynamic.code

import android.annotation.SuppressLint
import android.content.res.Configuration
import android.content.res.Resources
import android.util.TypedValue
import ag.virtues.dimens.library.DpQualifier
import ag.virtues.dimens.library.DpQualifierEntry
import ag.virtues.dimens.library.ScreenType
import ag.virtues.dimens.library.UiModeQualifierEntry
import ag.virtues.dimens.library.UiModeType

/**
 * [EN] Class for building dynamic dimensions that allow base Dp customization
 * via screen qualifiers (`.screen()`). The final value is scaled by the screen.
 * Compatible with the View System (XML layouts).
 *
 * [PT] Classe para construir dimensões dinâmicas que permitem customização do DP base
 * via qualificadores de tela (`.screen()`). O valor final é escalado pela tela.
 * Compatível com View System (XML layouts).
 */
class VirtuesDynamic(
    private val initialBaseDp: Float,
    private var ignoreMultiViewAdjustment: Boolean = false
) {
    /**
     * [EN] Map to store custom Dp values (Priority 3).
     * [PT] Mapa para armazenar valores Dp customizados (Prioridade 3).
     */
    private var customDpMap: MutableMap<DpQualifierEntry, Float> = mutableMapOf()

    /**
     * [EN] Map for custom Dp values by UiModeType (Priority 2).
     * [PT] Mapa para valores Dp customizados por UiModeType (Prioridade 2).
     */
    private var customUiModeMap: MutableMap<UiModeType, Float> = mutableMapOf()

    /**
     * [EN] Map for custom Dp values by INTERSECTION (UiMode + DpQualifier) (Priority 1).
     * [PT] Mapa para valores Dp customizados por INTERSEÇÃO (UiMode + DpQualifier) (Prioridade 1).
     */
    private var customIntersectionMap: MutableMap<UiModeQualifierEntry, Float> = mutableMapOf()

    /**
     * [EN] Defines the screen dimension to be used as a base (HIGHEST or LOWEST).
     * [PT] Define a dimensão da tela a ser usada como base (HIGHEST ou LOWEST).
     */
    private var screenType: ScreenType = ScreenType.LOWEST

    /**
     * [EN] Sets a custom dimension value for a specific UI mode.
     * @param type The UI mode (`UiModeType`).
     * @param customValue The custom dimension value in Dp.
     * @return The `VirtuesDynamic` instance for chaining.
     *
     * [PT] Define um valor de dimensão customizado para um modo de UI específico.
     * @param type O modo de UI (`UiModeType`).
     * @param customValue O valor de dimensão customizado em Dp.
     * @return A instância `VirtuesDynamic` para encadeamento.
     */
    fun screen(type: UiModeType, customValue: Float): VirtuesDynamic {
        customUiModeMap[type] = customValue
        return this
    }

    fun screen(type: UiModeType, customValue: Int): VirtuesDynamic {
        customUiModeMap[type] = customValue.toFloat()
        return this
    }

    /**
     * [EN] Sets a custom dimension for the intersection of a UI mode and a screen qualifier.
     * @param uiModeType The UI mode (`UiModeType`).
     * @param qualifierType The qualifier type (`DpQualifier`).
     * @param qualifierValue The qualifier value (e.g., 600 for sw600dp).
     * @param customValue The custom dimension value in Dp.
     * @return The `VirtuesDynamic` instance for chaining.
     *
     * [PT] Define uma dimensão customizada para a interseção de um modo de UI e um qualificador de tela.
     * @param uiModeType O modo de UI (`UiModeType`).
     * @param qualifierType O tipo de qualificador (`DpQualifier`).
     * @param qualifierValue O valor do qualificador (ex: 600 para sw600dp).
     * @param customValue O valor de dimensão customizado em Dp.
     * @return A instância `VirtuesDynamic` para encadeamento.
     */
    fun screen(
        uiModeType: UiModeType,
        qualifierType: DpQualifier,
        qualifierValue: Int,
        customValue: Float
    ): VirtuesDynamic {
        val key = UiModeQualifierEntry(
            uiModeType = uiModeType,
            dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue)
        )
        customIntersectionMap[key] = customValue
        return this
    }

    fun screen(
        uiModeType: UiModeType,
        qualifierType: DpQualifier,
        qualifierValue: Int,
        customValue: Int
    ): VirtuesDynamic {
        val key = UiModeQualifierEntry(
            uiModeType = uiModeType,
            dpQualifierEntry = DpQualifierEntry(qualifierType, qualifierValue)
        )
        customIntersectionMap[key] = customValue.toFloat()
        return this
    }

    /**
     * [EN] Sets a custom dimension value for a specific screen qualifier.
     * @param type The qualifier type (`DpQualifier`).
     * @param value The qualifier value (e.g., 600 for sw600dp).
     * @param customValue The custom dimension value in Dp.
     * @return The `VirtuesDynamic` instance for chaining.
     *
     * [PT] Define um valor de dimensão customizado para um qualificador de tela específico.
     * @param type O tipo de qualificador (`DpQualifier`).
     * @param value O valor do qualificador (ex: 600 para sw600dp).
     * @param customValue O valor da dimensão customizada em Dp.
     * @return A instância `VirtuesDynamic` para encadeamento.
     */
    fun screen(type: DpQualifier, value: Int, customValue: Int): VirtuesDynamic {
        customDpMap[DpQualifierEntry(type, value)] = customValue.toFloat()
        return this
    }

    fun screen(type: DpQualifier, value: Int, customValue: Float): VirtuesDynamic {
        customDpMap[DpQualifierEntry(type, value)] = customValue
        return this
    }

    /**
     * [EN] Sets the screen dimension type (LOWEST or HIGHEST) to be used as the base for adjustments.
     * @param type The screen dimension type.
     * @return The `VirtuesDynamic` instance for chaining.
     *
     * [PT] Define o tipo de dimensão da tela (LOWEST ou HIGHEST) a ser usado como base para os ajustes.
     * @param type O tipo de dimensão da tela.
     * @return A instância `VirtuesDynamic` para encadeamento.
     */
    fun type(type: ScreenType): VirtuesDynamic {
        screenType = type
        return this
    }

    /**
     * [EN] Ignores adjustments when the app is in multi-window mode.
     * @param ignore If true, adjustments are ignored in multi-window mode.
     * @return The `VirtuesDynamic` instance for chaining.
     *
     * [PT] Ignora os ajustes quando o aplicativo está em modo multi-janela.
     * @param ignore Se verdadeiro, os ajustes são ignorados no modo multi-janela.
     * @return A instância `VirtuesDynamic` para encadeamento.
     */
    fun multiViewAdjustment(ignore: Boolean = true): VirtuesDynamic {
        ignoreMultiViewAdjustment = ignore
        return this
    }

    /**
     * [EN] Resolves the base Dp value to be adjusted by applying the customization logic
     * (Intersection > UiMode > DpQualifier).
     *
     * @param configuration The current screen configuration.
     * @return The base Dp value (unadjusted for the screen) to be used in the calculation.
     *
     * [PT] Resolve o valor Dp base a ser ajustado, aplicando a lógica de customização
     * (Interseção > UiMode > DpQualifier).
     *
     * @param configuration A configuração da tela atual.
     * @return O valor Dp base (não ajustado pela tela) a ser usado no cálculo.
     */
    @SuppressLint("ConfigurationScreenWidthHeight")
    private fun resolveFinalBaseDp(configuration: Configuration): Float {
        val smallestWidthDp = configuration.smallestScreenWidthDp.toFloat()
        val currentScreenWidthDp = configuration.screenWidthDp.toFloat()
        val currentScreenHeightDp = configuration.screenHeightDp.toFloat()
        val currentUiModeType = UiModeType.fromConfiguration(configuration.uiMode)

        var dpToAdjust = initialBaseDp
        var foundCustomDp: Float?

        // Priority 1: Intersection (UiMode + DpQualifier)
        val sortedIntersectionQualifiers = customIntersectionMap.entries.toList()
            .sortedByDescending { it.key.dpQualifierEntry.value }

        foundCustomDp = sortedIntersectionQualifiers.firstOrNull { (key, _) ->
            key.uiModeType == currentUiModeType && VirtuesAdjustmentFactors.resolveIntersectionCondition(
                entry = key.dpQualifierEntry,
                smallestWidthDp = smallestWidthDp,
                currentScreenWidthDp = currentScreenWidthDp,
                currentScreenHeightDp = currentScreenHeightDp
            )
        }?.value

        if (foundCustomDp != null) {
            dpToAdjust = foundCustomDp
        } else {
            // Priority 2: UI Mode (UiModeType only)
            foundCustomDp = customUiModeMap[currentUiModeType]

            if (foundCustomDp != null) {
                dpToAdjust = foundCustomDp
            } else {
                // Priority 3: Dp Qualifier (SW, H, W only)
                dpToAdjust = resolveQualifierDpFloat(
                    customDpMap = customDpMap,
                    smallestWidthDp = smallestWidthDp,
                    currentScreenWidthDp = currentScreenWidthDp,
                    currentScreenHeightDp = currentScreenHeightDp,
                    initialBaseDp = initialBaseDp
                )
            }
        }
        return dpToAdjust
    }

    /**
     * [EN] Auxiliary version of `resolveQualifierDp` that works with Float instead of Dp.
     * [PT] Versão auxiliar do `resolveQualifierDp` que trabalha com Float ao invés de Dp.
     */
    private fun resolveQualifierDpFloat(
        customDpMap: Map<DpQualifierEntry, Float>,
        smallestWidthDp: Float,
        currentScreenWidthDp: Float,
        currentScreenHeightDp: Float,
        initialBaseDp: Float
    ): Float {
        val sortedQualifiers = customDpMap.entries.toList()
            .sortedByDescending { it.key.value }

        val foundCustomDp = sortedQualifiers.firstOrNull { (entry, _) ->
            VirtuesAdjustmentFactors.resolveIntersectionCondition(
                entry = entry,
                smallestWidthDp = smallestWidthDp,
                currentScreenWidthDp = currentScreenWidthDp,
                currentScreenHeightDp = currentScreenHeightDp
            )
        }?.value

        return foundCustomDp ?: initialBaseDp
    }

    /**
     * [EN] Performs the final dynamic dimension calculation.
     *
     * @param configuration The current screen configuration.
     * @return The dynamically adjusted Dp value as a **Float** (not converted to PX/SP).
     *
     * [PT] Executa o cálculo final da dimensão dinâmica.
     *
     * @param configuration A Configuration da tela atual.
     * @return O valor **Float** em Dp ajustado dinamicamente (não convertido para PX/SP).
     */
    @SuppressLint("ConfigurationScreenWidthHeight")
    fun calculateAdjustedValue(configuration: Configuration): Float {
        val dpToAdjust = resolveFinalBaseDp(configuration)

        var isMultiWindow = false

        if (ignoreMultiViewAdjustment) {
            val smallestWidthDp = configuration.smallestScreenWidthDp.toFloat()
            val currentScreenWidthDp = configuration.screenWidthDp.toFloat()
            val isLayoutSplit = configuration.screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK != Configuration.SCREENLAYOUT_SIZE_MASK
            val isSmallDifference = (smallestWidthDp - currentScreenWidthDp) < (smallestWidthDp * 0.1)
            isMultiWindow = isLayoutSplit && !isSmallDifference
        }

        val shouldIgnoreAdjustment = ignoreMultiViewAdjustment && isMultiWindow

        if (shouldIgnoreAdjustment) {
            // Returns the base Dp without dynamic scaling
            return dpToAdjust
        }

        // The dynamic scaling percentage is: (Adjusted Base DP / Reference DP)
        val percentage = dpToAdjust / VirtuesAdjustmentFactors.BASE_WIDTH_DP

        // Screen dimension to use (LOWEST or HIGHEST)
        val dimensionToUse = when (screenType) {
            ScreenType.HIGHEST -> maxOf(
                configuration.screenWidthDp.toFloat(),
                configuration.screenHeightDp.toFloat()
            )
            ScreenType.LOWEST -> minOf(
                configuration.screenWidthDp.toFloat(),
                configuration.screenHeightDp.toFloat()
            )
        }

        // The final value is the percentage applied to the screen dimension
        return dimensionToUse * percentage
    }

    /**
     * [EN] Builds the dynamically adjusted Dp value and converts it to Pixels (Float).
     *
     * @param resources The Context's Resources.
     * @return The value in Pixels (PX) as a Float.
     *
     * [PT] Constrói o valor Dp ajustado dinamicamente e o converte para Pixels (Float).
     *
     * @param resources Os Resources do Context.
     * @return O valor em Pixels (PX) como Float.
     */
    fun toPx(resources: Resources): Float {
        val adjustedDpValue = calculateAdjustedValue(resources.configuration)
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, adjustedDpValue, resources.displayMetrics
        )
    }

    /**
     * [EN] Builds the dynamically adjusted Dp value and converts it to Pixels (Int).
     * Useful for setters that only accept Int.
     *
     * @param resources The Context's Resources.
     * @return The value in Pixels (PX) as an Int.
     *
     * [PT] Constrói o valor Dp ajustado dinamicamente e o converte para Pixels (Int).
     * Útil para setters que aceitam apenas Int.
     *
     * @param resources Os Resources do Context.
     * @return O valor em Pixels (PX) como Int.
     */
    fun toPxInt(resources: Resources): Int {
        return toPx(resources).toInt()
    }

    /**
     * [EN] Builds the dynamically adjusted Dp value and converts it to Scalable Pixels (SP) in Pixels (Float),
     * ignoring the system's font scale ('em' unit).
     *
     * @param resources The Context's Resources.
     * @return The value in Pixels (PX) corresponding to 'em', as a Float.
     *
     * [PT] Constrói o valor Dp ajustado dinamicamente e o converte para Scaleable Pixels (SP) em Pixels (Float),
     * ignorando a escala de fonte do sistema (unidade 'em').
     *
     * @param resources Os Resources do Context.
     * @return O valor em Pixels (PX) correspondente ao 'em', como Float.
     */
    fun toEm(resources: Resources): Float {
        val adjustedDpValue = calculateAdjustedValue(resources.configuration)
        val fontScale = resources.configuration.fontScale
        val dpValueWithoutFontScale = adjustedDpValue / fontScale
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_SP, dpValueWithoutFontScale, resources.displayMetrics
        )
    }

    /**
     * [EN] Builds the dynamically adjusted Dp value and converts it to Scalable Pixels (SP) in Pixels (Int),
     * ignoring the system's font scale ('em' unit).
     *
     * @param resources The Context's Resources.
     * @return The value in Pixels (PX) corresponding to 'em', as an Int.
     *
     * [PT] Constrói o valor Dp ajustado dinamicamente e o converte para Scaleable Pixels (SP) em Pixels (Int),
     * ignorando a escala de fonte do sistema (unidade 'em').
     *
     * @param resources Os Resources do Context.
     * @return O valor em Pixels (PX) correspondente ao 'em', como Int.
     */
    fun toEmInt(resources: Resources): Int {
        return toEm(resources).toInt()
    }

    /**
     * [EN] Builds the dynamically adjusted Dp value and converts it to Scalable Pixels (SP) in Pixels (Float).
     *
     * @param resources The Context's Resources.
     * @return The value in Pixels (PX) corresponding to SP, as a Float.
     *
     * [PT] Constrói o valor Dp ajustado dinamicamente e o converte para Scaleable Pixels (SP) em Pixels (Float).
     *
     * @param resources Os Resources do Context.
     * @return O valor em Pixels (PX) correspondente ao SP, como Float.
     */
    fun toSp(resources: Resources): Float {
        val adjustedDpValue = calculateAdjustedValue(resources.configuration)
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_SP, adjustedDpValue, resources.displayMetrics
        )
    }

    /**
     * [EN] Builds the dynamically adjusted Dp value and converts it to Scalable Pixels (SP) in Pixels (Int).
     *
     * @param resources The Context's Resources.
     * @return The value in Pixels (PX) corresponding to SP, as an Int.
     *
     * [PT] Constrói o valor Dp ajustado dinamicamente e o converte para Scaleable Pixels (SP) em Pixels (Int).
     *
     * @param resources Os Resources do Context.
     * @return O valor em Pixels (PX) correspondente ao SP, como Int.
     */
    fun toSpInt(resources: Resources): Int {
        return toSp(resources).toInt()
    }

    /**
     * [EN] Returns the dynamically adjusted Dp value (in Dp, not converted to PX).
     *
     * @param resources The Context's Resources.
     * @return The value in Dp as a Float.
     *
     * [PT] Retorna o valor Dp ajustado dinamicamente (em Dp, não convertido para PX).
     *
     * @param resources Os Resources do Context.
     * @return O valor em Dp como Float.
     */
    fun toDp(resources: Resources): Float {
        return calculateAdjustedValue(resources.configuration)
    }

    /**
     * [EN] Returns the dynamically adjusted Dp value (in Dp, not converted to PX) as an Int.
     *
     * @param resources The Context's Resources.
     * @return The value in Dp as an Int.
     *
     * [PT] Retorna o valor Dp ajustado dinamicamente (em Dp, não convertido para PX) como Int.
     *
     * @param resources Os Resources do Context.
     * @return O valor em Dp como Int.
     */
    fun toDpInt(resources: Resources): Int {
        return calculateAdjustedValue(resources.configuration).toInt()
    }
}

File: iOS/Sources/VirtuesDimens/AppDimensDynamic.swift
/**
 * Author & Developer: Jean Bodenberg
 * GIT: https://github.com/www-virtues-ag/appdimens.git
 * Date: 2025-01-15
 *
 * Library: VirtuesDimens iOS
 *
 * Description:
 * The VirtuesDimens library is a dimension management system that automatically
 * adjusts Dp, Sp, and Px values in a responsive and mathematically refined way,
 * ensuring layout consistency across any screen size or ratio.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation
import UIKit

/**
 * [EN] Class for building dynamic dimensions that allow base value customization
 * via screen qualifiers. The final value is scaled by the screen.
 * Compatible with both UIKit and SwiftUI.
 * [PT] Classe para construir dimensões dinâmicas que permitem customização do valor base
 * via qualificadores de tela. O valor final é escalado pela tela.
 * Compatível com UIKit e SwiftUI.
 */
public class VirtuesDimensDynamic {
    
    // MARK: - Properties
    
    private let initialBaseValue: CGFloat
    private var ignoreMultiWindowAdjustment: Bool = false
    private var screenType: ScreenType = .lowest
    
    /// [EN] Map to store custom values based on device type (Priority 2).
    /// [PT] Mapa para armazenar valores customizados com base no tipo de dispositivo (Prioridade 2).
    private var customDeviceTypeMap: [DeviceType: CGFloat] = [:]
    
    /// [EN] Map to store custom values based on screen qualifier (Priority 3).
    /// [PT] Mapa para armazenar valores customizados com base no qualificador de tela (Prioridade 3).
    private var customScreenQualifierMap: [ScreenQualifierEntry: CGFloat] = [:]
    
    // MARK: - Initialization
    
    /**
     * [EN] Initializes the VirtuesDimensDynamic with a base value.
     * [PT] Inicializa o VirtuesDimensDynamic com um valor base.
     */
    public init(_ initialValue: CGFloat, ignoreMultiWindowAdjustment: Bool = false) {
        self.initialBaseValue = initialValue
        self.ignoreMultiWindowAdjustment = ignoreMultiWindowAdjustment
    }
    
    // MARK: - Configuration Methods
    
    /**
     * [EN] Sets a custom dimension value for a specific device type.
     * @param type The device type.
     * @param customValue The custom dimension value in points.
     * @return The VirtuesDimensDynamic instance for chaining.
     * [PT] Define um valor de dimensão customizado para um tipo de dispositivo específico.
     * @param type O tipo de dispositivo.
     * @param customValue O valor de dimensão customizado em pontos.
     * @return A instância VirtuesDimensDynamic para encadeamento.
     */
    @discardableResult
    public func screen(_ type: DeviceType, _ customValue: CGFloat) -> VirtuesDimensDynamic {
        customDeviceTypeMap[type] = customValue
        return self
    }
    
    /**
     * [EN] Sets a custom dimension value for a specific device type and screen size.
     * @param deviceType The device type.
     * @param screenSize The minimum screen size for this qualifier.
     * @param customValue The custom dimension value in points.
     * @return The VirtuesDimensDynamic instance for chaining.
     * [PT] Define um valor de dimensão customizado para um tipo de dispositivo e tamanho de tela específicos.
     * @param deviceType O tipo de dispositivo.
     * @param screenSize O tamanho mínimo de tela para este qualificador.
     * @param customValue O valor de dimensão customizado em pontos.
     * @return A instância VirtuesDimensDynamic para encadeamento.
     */
    @discardableResult
    public func screen(_ deviceType: DeviceType, _ screenSize: CGFloat, _ customValue: CGFloat) -> VirtuesDimensDynamic {
        let entry = ScreenQualifierEntry(deviceType: deviceType, screenSize: screenSize)
        customScreenQualifierMap[entry] = customValue
        return self
    }
    
    /**
     * [EN] Sets the screen dimension type (LOWEST or HIGHEST) to be used as the base for adjustments.
     * @param type The screen dimension type.
     * @return The VirtuesDimensDynamic instance for chaining.
     * [PT] Define o tipo de dimensão da tela (LOWEST ou HIGHEST) a ser usado como base para os ajustes.
     * @param type O tipo de dimensão da tela.
     * @return A instância VirtuesDimensDynamic para encadeamento.
     */
    @discardableResult
    public func type(_ type: ScreenType) -> VirtuesDimensDynamic {
        screenType = type
        return self
    }
    
    /**
     * [EN] Ignores adjustments when the app is in multi-window mode.
     * @param ignore If true, adjustments are ignored in multi-window mode.
     * @return The VirtuesDimensDynamic instance for chaining.
     * [PT] Ignora os ajustes quando o aplicativo está em modo multi-janela.
     * @param ignore Se verdadeiro, os ajustes são ignorados no modo multi-janela.
     * @return A instância VirtuesDimensDynamic para encadeamento.
     */
    @discardableResult
    public func multiWindowAdjustment(ignore: Bool = true) -> VirtuesDimensDynamic {
        ignoreMultiWindowAdjustment = ignore
        return self
    }
    
    // MARK: - Value Resolution
    
    /**
     * [EN] Resolves the base value to be adjusted by applying the customization logic
     * (Device Type > Screen Qualifier > Initial Value).
     * [PT] Resolve o valor base a ser ajustado, aplicando a lógica de customização
     * (Tipo de Dispositivo > Qualificador de Tela > Valor Inicial).
     */
    private func resolveFinalBaseValue() -> CGFloat {
        let currentDeviceType = DeviceType.current()
        let (screenWidth, screenHeight) = VirtuesDimensAdjustmentFactors.getCurrentScreenDimensions()
        let currentScreenSize = min(screenWidth, screenHeight)
        
        // Priority 1: Device Type
        if let deviceTypeValue = customDeviceTypeMap[currentDeviceType] {
            return deviceTypeValue
        }
        
        // Priority 2: Screen Qualifier
        let qualifierValue = VirtuesDimensAdjustmentFactors.resolveQualifierValue(
            customMap: customScreenQualifierMap,
            currentDeviceType: currentDeviceType,
            currentScreenSize: currentScreenSize,
            initialValue: initialBaseValue
        )
        
        return qualifierValue
    }
    
    /**
     * [EN] Performs the final dynamic dimension calculation.
     * @return The dynamically adjusted value as a CGFloat (not converted to pixels).
     * [PT] Executa o cálculo final da dimensão dinâmica.
     * @return O valor ajustado dinamicamente como CGFloat (não convertido para pixels).
     */
    public func calculateAdjustedValue() -> CGFloat {
        let valueToAdjust = resolveFinalBaseValue()
        
        let shouldIgnoreAdjustment = ignoreMultiWindowAdjustment && VirtuesDimensAdjustmentFactors.isMultiWindowMode()
        
        if shouldIgnoreAdjustment {
            // Returns the base value without dynamic scaling
            return valueToAdjust
        }
        
        // The dynamic scaling percentage is: (Adjusted Base Value / Reference Value)
        let percentage = valueToAdjust / VirtuesDimensAdjustmentFactors.BASE_WIDTH_PT
        
        // Screen dimension to use (LOWEST or HIGHEST)
        let (screenWidth, screenHeight) = VirtuesDimensAdjustmentFactors.getCurrentScreenDimensions()
        let dimensionToUse = screenType == .highest ? 
            max(screenWidth, screenHeight) : 
            min(screenWidth, screenHeight)
        
        // The final value is the percentage applied to the screen dimension
        return dimensionToUse * percentage
    }
    
    // MARK: - Conversion Methods
    
    /**
     * [EN] Returns the dynamically adjusted value in points.
     * @return The value in points as a CGFloat.
     * [PT] Retorna o valor ajustado dinamicamente em pontos.
     * @return O valor em pontos como CGFloat.
     */
    public func toPoints() -> CGFloat {
        return calculateAdjustedValue()
    }
    
    /**
     * [EN] Returns the dynamically adjusted value in pixels.
     * @return The value in pixels as a CGFloat.
     * [PT] Retorna o valor ajustado dinamicamente em pixels.
     * @return O valor em pixels como CGFloat.
     */
    public func toPixels() -> CGFloat {
        let adjustedValue = calculateAdjustedValue()
        return VirtuesDimensAdjustmentFactors.pointsToPixels(adjustedValue)
    }
    
    /**
     * [EN] Returns the dynamically adjusted value as an integer in points.
     * @return The value in points as an Int.
     * [PT] Retorna o valor ajustado dinamicamente como inteiro em pontos.
     * @return O valor em pontos como Int.
     */
    public func toPointsInt() -> Int {
        return Int(calculateAdjustedValue())
    }
    
    /**
     * [EN] Returns the dynamically adjusted value as an integer in pixels.
     * @return The value in pixels as an Int.
     * [PT] Retorna o valor ajustado dinamicamente como inteiro em pixels.
     * @return O valor em pixels como Int.
     */
    public func toPixelsInt() -> Int {
        return Int(toPixels())
    }
}
